### DISCUSSION TITLE
Controversial topic - beautiful issue and PR buffers versus editable buffer

### DISCUSSION THREAD

**Discussion by weilbith** [pwntester/octo.nvim#112] [Ideas]
I was thinking about your complains that the buffers do not look too awesome and this could be improved. After all any improvement or at least many suffer by the issue that the comment bodies and other editable stuff must be "original" text that can't be formatted or anything. Everything could be changed (design) wise, if the editable feature could get removed/outsourced. This approach would produce an incredible beautiful buffer (still having some extmarks) and then habe a mapping to edit the object under the cursor. The user would get a quick feedback if can actually edit this part (maybe it's not his comment and he is not an admin). Then the original unformatted content opens in a pop-up window. The user can edit everything and then save and close it. This will do the update (should auto-update the origin buffer content).

I think this feature of having an editable issue buffer is pretty cool. But I kinda seems like it is limiting all other possible features. With an approach like the above it would also produce a more generic approach. Means the user says: edit under the cursor. This could be the floating window to edit a text, but also for other fields like adding a reviewer, a label (as already is), a reaction etc pp. It would open the plugin for many new pretty illustrations **and** cool usability features.

For the beautified version I'm also thinking about highlighting the users own comment differently (as the WebUI does). A vertical timeline. Boxes around contents, ... the folds should remain. The sign column wrappers would be no more necessary. Different indent level would possible. Limiting comment width is possible. Filetype highlight injection for comments. Having a proper comment buffer with Markdown mistype so the user gets all the features he has setup (e.g. emoji completion, render preview, concealing, snippets, ...). No "cumbersome" visual updates of the virtual text and signs according to in-buffer changes. And many actions in these buffers are already commands/mappings to change/edit them. So after all it is only a small part that needs to adopt and this would make it completely consistent. Also if the feature to add and edit comments during a review will be done in a floating window, this is even more consistent in the issue/PR buffers.
The extmarks would be used to determine which object should be edited. After all this should actually not too much if a code change. Some things can be removed and get re-used for the new edit process. The pR shouldn't be big. And afterwards all the new improvement become easily possible.

This text is quite chaotic. I hope it is still enough for a good discussion.
After all I don't want to critize the plugins approach. But after after our recent discussions I was just thinking about stuff and this was the central question that affects everything (or many things). So I thought it is worth to discuss it. Or you just say: no, I insist on the in-buffer edit feature. That would be absolute fine as well!
I can't say it often enough: thanks for this great plugin! :pray:

**Comment by weilbith** [comment]
PS: I finally started to cruise around in the code base to get used to it and understand the details. So I hope I can do some PRs in near future. :v:

**Comment by weilbith** [comment]
Just came into my mind: how could would it be to not have "complicated" commands or mappings that need to be remembered, but using the insert enter autocmd to detect/initiate the change that then will trigger the action handler (e.g. open the float window to edit comment). Ofc we would always immediately leave insert mode and detect the current cursor position, using the extmarks to detect the object to change. This would also flatten the huddle of this "breaking" change for users who are already used to the edit-the-issue-buffer feature.

**Comment by pwntester** [comment]
Thanks for starting the discussion, I think there is a lot of good ideas in there but I really like the in-buffer editing. I often find myself changing buffers to copy/paste text or code from somewhere else while the comment is still being edited. I feel like using floats is more like a one-shot chance to write the comment and you lose the ability to move around. You could always write some part of the comment in the float, then exit it (saving to GH), copy something from somewhere else and editing again the comment, but I think that doesnt flow well.

Having said that, I already thought about the idea of treating the octo buffers as non-modifiable (entering insert mode will leave back to normal immediately) except for pieces (such as comments) that you are allowed to edit. This would make virtual text unnecessary, or at least prevent users from writing on those lines. Also, it allow us to have to modes: view and edit. In view mode, we could format the text as we want (indents, line width, vertical guidelines, boxes, ...), but on edit mode, the extmarked region would be replaced by the unformatted text to allow users to freely modify it. Back to view mode, we can re-apply  formatting as needed. This would give us the best of the two worlds I think

> Means the user says: edit under the cursor. This could be the floating window to edit a text, but also for other fields like adding a reviewer, a label (as already is), a reaction etc pp.

I think this can also be applied to the proposed mixed solution. Entering insert mode on a label would open the popup window to pick new labels from, etc ...

> The sign column wrappers would be no more necessary. 

In this mixed mode, it would still be necessary to show that the comment is being edited. Or we can change the linehl to a different background color

**Reply by pwntester** [reply]
I already implemented a control to prevent editing of non-editable regions in master.
Also, I started a PoC implementation of the render vs edit modes at #113

**Reply by weilbith** [reply]
I'll respond everything soon and will also create the asked for issues.
But as a quick response: `master` is currently broken for me when I want to open a PR:
```vim
E5108: Error executing lua ...im_test_site/pack/packer/opt/octo.nvim/lua/octo/menu.lua:80: Vim(lua):E5108: Error executing lua [string ":lua"]:1: attempt to call field 'get_pull_request' (a nil value)
```

**Reply by pwntester** [reply]
oops, forgot to update one dynamically constructed location. should be ok now

**Comment by weilbith** [comment]
As there was some off-topic conversation below the last comment I just continue on this level here.

---

First I created the both issues you have asked for: #114 #115

---

Alright. I understand you. Thanks for making a clear statement. :+1: 
Thinking about this hybrid solution, I wonder/worry how the transition would look like? The editable view is pretty much restricted. Depending on how fancy the view mode becomes, how would this affect the user experience if everything moves, shifts, reshapes etc. pp. Do you already have some more detailed ideas for that? Maybe a quite generic description if possible. :grimacing: 

For the insert mode edit extensions for label etc. I think #101 is already addressing them, right?

**Reply by pwntester** [reply]
It seems like the hybrid approach would not be possible after all. I can hijack the insert mode, but I havent found a way to prevent or control modifications from normal mode (eg: `dd`) so you may delete lines from a "view" mode rendered comment for example

Another solution may be to use floats but do not sync immediatly with GitHub. This way the user can popup the comment and edit it repeatly until its ok to submit it to GH. We need to keep the sign column indicators and the `BufWriteCmd` hook to sync with GH as we are doing now

**Reply by weilbith** [reply]
Sorry to hear that the hybrid does not work. :slightly_frowning_face: 

Sounds pretty cool. Thinking about it it was actually never my intention to execute changes immediately after the float closes.
How do you intend to store the intermediate state of unsaved changes? Just update the GH objects in the buffer variables and let the render function just re-render everything after the float closes? :thinking: 
Well, actually that's likely already in place (except the re-render). :joy: 
Why do you need to keep the sign column indicators? :thinking: The file is changed. The user should see this usually. And if he deletes a buffer he should get asked if he wants to save. If he "disables" this default behavior we should not care? Or we just listed to the `BufDelete` event to assist the user? Anyway we should still have visual highlights for changed content in the issue buffer. But that must not be the signs, but we are then open to do anything we want to. Right?

**Reply by pwntester** [reply]
> Well, actually that's likely already in place (except the re-render)

Yep, it shouldn't be difficult, everything is already there

> Anyway we should still have visual highlights for changed content in the issue buffer. But that must not be the signs, but we are then open to do anything we want to. Right?

True, we could add that indicator in the "indentation" part of the comment or as a red box around the comment and get rid of the sign column altogether.

**Reply by pwntester** [reply]
still not 100 percent sure about removing in-buffer edit. It works so great for me ....

**Reply by weilbith** [reply]
I can understand. Maybe just think a bit longer about it and focus on other features which are quite independent of this (at least their stubs). Maybe over time their comes a better idea or rather a solution for the problem. After all you have already dozen of visual improvements in their and it is not a real feature, rather eye candy/better readability (which its value is not to underestimate though).

**Comment by williambdean** [comment]
Closing due to updated nature of the discussion. If there are some specific changes that would like to be done, I'm happy to reopen this or create specific issues for these! Hopefully the buffers are looking really nice these days!
